client/config.go

```diff
@@ -1,131 +1,164 @@
 
-	"github.com/cosmos/cosmos-sdk/types"
-	"github.com/mitchellh/go-homedir"
 	"github.com/pelletier/go-toml"
 	"github.com/spf13/cobra"
+	"github.com/spf13/viper"
+)
+
+const (
+	flagGet  = "get"
+	flagList = "list"
 )
 
-type cliConfig struct {
-	Home      string `toml:"home"`
-	ChainID   string `toml:"chain_id"`
-	TrustNode bool   `toml:"trust_node"`
-	Output    string `toml:"output"`
-	Node      string `toml:"node"`
-	Trace     bool   `toml:"trace"`
+var configDefaults map[string]string
+
+func init() {
+	configDefaults = map[string]string{
+		"chain_id": "",
+		"output":   "text",
+		"node":     "tcp://localhost:26657",
+	}
 }
 
 // ConfigCmd returns a CLI command to interactively create a
 // Gaia CLI config file.
 func ConfigCmd() *cobra.Command {
-	cfg := &cobra.Command{
-		Use:   "config",
-		Short: "Interactively creates a Gaia CLI config file",
+	cmd := &cobra.Command{
+		Use:   "config <key> [value]",
+		Short: "Create or query a Gaia CLI configuration file",
 		RunE:  runConfigCmd,
+		Args:  cobra.MaximumNArgs(2),
 	}
 
-	return cfg
+	cmd.Flags().String(cli.HomeFlag, app.DefaultCLIHome,
+		"set client's home directory for configuration")
+	cmd.Flags().Bool(flagGet, false,
+		"print configuration value or its default if unset")
+	cmd.Flags().BoolP(flagList, "l", false,
+		"list all variables set in the config file, along with their values")
+	return cmd
 }
 
 func runConfigCmd(cmd *cobra.Command, args []string) error {
-	home, err := homedir.Dir()
+	cfgFile, err := ensureConfFile(viper.GetString(cli.HomeFlag))
 	if err != nil {
 		return err
 	}
 
-	stdin := BufferStdin()
+	getAction, listAction := viper.GetBool(flagGet), viper.GetBool(flagList)
+	// Args validation
+	if listAction && getAction {
+		return fmt.Errorf("only one action at a time")
+	}
 
-	gaiaCLIHome, err := handleGaiaCLIHome(home, stdin)
-	if err != nil {
-		return err
+	if (listAction && len(args) != 0) ||
+		(getAction && len(args) != 1) || len(args) != 2 {
+		return fmt.Errorf("wrong number of arguments")
 	}
 
-	node, err := handleNode(stdin)
+	// Load configuration
+	tree, err := loadConfigFile(cfgFile)
 	if err != nil {
 		return err
 	}
 
-	trustNode, err := handleTrustNode(stdin)
-	if err != nil {
-		return err
+	// List action
+	if listAction {
+		s, err := tree.ToTomlString()
+		if err != nil {
+			return err
+		}
+		fmt.Print(s)
+		return nil
 	}
 
-	chainID, err := types.DefaultChainID()
+	key := args[0]
+	// Get value action
+	if getAction {
+		switch key {
+		case "chain_id", "output", "node":
+			defaultValue := configDefaults[key]
+			fmt.Println(tree.GetDefault(key, defaultValue).(string))
+		case "trace", "trust_node":
+			fmt.Println(tree.GetDefault(key, false).(bool))
+		default:
+			return errUnknownConfigKey(key)
+		}
+		return nil
+	}
 
-	if err != nil {
-		fmt.Println("Couldn't populate ChainID, so using an empty one.")
+	// Set value action
+	value := args[1]
+	switch key {
+	case "chain_id", "output", "node":
```

Comment:
Is there a way we can avoid hardcoding these values? Where is the `config` struct defined?

Reply:
there is no config struct defined anywhere

Reply:
Oh, should we define one, like we do for the Tendermint configs? It would be a bit clearer perhaps.

Reply:
I experienced trouble in marshaling/unmarshaling structs. Furthermore, doing it this way allows us to set default values.

Reply:
I think we can do that in either case - Tendermint certainly sets default values.

Reply:
If we don't set default values, the other gaiacli commands will override them. I don't think it's a big deal then

Reply:
However, pelletier got issues with TOML marshaling/unmarshalig, whilst the toml.Tree works pretty well

Reply:
OK, nbd for now.